astropy/cosmology/core.py

```diff
@@ -375,11 +388,70 @@ class FlatCosmologyMixin(metaclass=abc.ABCMeta):
     but ``FlatLambdaCDM`` **will** be flat.
     """
 
+    def __init_subclass__(cls: Type[_FlatCT]) -> None:
+        super().__init_subclass__()
+
+        # Determine the non-flat class.
+        # This will raise a TypeError if the MRO is inconsistent.
+        cls._nonflat_cls_
+
+    # ===============================================================
+
+    @classmethod  # TODO! make metaclass-method
+    def _get_nonflat_cls(cls: Type[_FlatCT], kls: Optional[Type[_CT]]=None) -> Type[_CT]:
+        """Find the corresponding non-flat class.
+
+        The class' bases are searched recursively.
+
+        Parameters
+        ----------
+        kls : :class:`astropy.cosmology.Cosmology` class or None, optional
+            If `None` (default) this class is searched instead of `kls`.
+
+        Raises
+        ------
+        TypeError
+            If more than one non-flat class is found at the same level of the
+            inheritance. This is similar to the error normally raised by
+            python for an inconsistent MRO.
+
+        Returns
+        -------
+        type
+            A Cosmology class that is not a FlatCosmologyMixin subclass.
+        """
+        _kls = cls if kls is None else kls  # Which class to search.
+
+        # Find non-flat classes
+        nonflat: List[Type[_CT]]
+        nonflat = [b for b in _kls.__bases__ if not issubclass(b, FlatCosmologyMixin)]
```

Comment:
The danger of switching to `set` is that it is not ordered, so if anything ever goes wrong it would be hard to reproduce. 
The list comprehension will produce the same results every time.

Reply:
If we do a breadth-first search the problems of `set` should be ameliorated.

Reply:
Sets are not ordered, but they are guaranteed not to have duplicate elements. But it could be that this property is not actually useful, in which case using a `list` would be fine. I still think the `issubclass(b, Cosmology)` check would be useful though.

Reply:
> I still think the issubclass(b, Cosmology) check would be useful though.

Agreed.